code/obj/submachine/robotics.dm

```diff
@@ -271,26 +271,63 @@
+	var/opmode = 1 // operating mode. 1 will be pickup and 2 will be interact (drop, attack)
+
+	attack_self(var/mob/user as mob) // Switch our operating mode here
+		switch (src.opmode)
+			if (1)
+				src.opmode = 2
+				user.show_text("Operating mode set to Dropoff / Interact.", "red")
+			if (2)
+				src.opmode = 1
+				user.show_text("Operating mode set to Pick-up", "green")
+
+	afterattack(atom/A, mob/user as mob) // I have no idea what I am doing but let's go for it.
+		if (!A) return // I assume this is so we don't try and interact with a non-existent atom.
+		// Time to consider our operating mode.
+		if (BOUNDS_DIST(get_turf(src), get_turf(A)) > 0) // If it is too far away do nothing.
+			return
+		if (src.opmode == 1) // We are in pickup mode, no reason to do anything.
+			return
+		else // We are in dropoff or interact mode.
+			if (length(src.contents) == 0) // We have nothing with which to interact/dropoff with.
+				boutput(user, SPAN_ALERT("You have nothing stored!"))
 				return
-			var/selection = tgui_input_list(user, "What do you want to drop?", "Atmospherics Transporter", src.contents)
-			if(!selection) return
-			if (istype(selection, /obj/machinery/fluid_canister))
-				var/obj/machinery/fluid_canister/S = selection
-				S.set_loc(get_turf(user.loc))
-				S.contained = 0
-			else if (istype(selection, /obj/machinery/portable_atmospherics))
-				var/obj/machinery/portable_atmospherics/S = selection
-				S.set_loc(get_turf(user.loc))
-				S.contained = 0
-			else return //no sparks for unintended items
-			elecflash(user)
-
+			else
+				//Let us determine if we are dealing with a floor/space or an object/machine.
+				if (istype(A, /turf/simulated/floor) || istype(A, /obj/lattice) || istype(A, /turf/space)) // Floor/space it is.
+					if (A.density) // really hoping this just check if a given turf is impassable
+						boutput(user, SPAN_ALERT("Not enough space here!"))
+						return
+					var/selection = tgui_input_list(user, "What do you want to drop?", "Atmospherics Transporter", src.contents) // Let us choose our thing to drop.
+					if(!selection)
+						return
+					if (istype(selection, /obj/machinery/fluid_canister))
+						var/obj/machinery/fluid_canister/S = selection
+						S.set_loc(A)
+						S.contained = 0
+					else if (istype(selection, /obj/machinery/portable_atmospherics))
+						var/obj/machinery/portable_atmospherics/S = selection
+						S.set_loc(A)
+						S.contained = 0
+					else if (istype(selection, /obj/item/tank))
+						var/obj/item/tank/S = selection
+						S.set_loc(A)
+					else return //no sparks for unintended items
+					elecflash(user)
+					return
+				if (istype(A, /obj/machinery/portable_atmospherics) || istype(A, /obj/machinery/power/collector_array)) // It is a portable atmos machine, cannister, or a singulo collector array.
+					var/selection = tgui_input_list(user, "What do you want to use?", "Atmospherics Transporter", src.contents) // Let us choose an item to smack it with.
```

Comment:
you could pre-filter a list of just the relevant tank contents, check its length to prevent no-op tgui popups, and  *also* guarantee that a tank was selected (if one exists).

Reply:
I could, but I do not know how to do this. Is it a deal breaker to not filter out non-tanks? We could file this under "expanding the player's freedom to make the wrong choice".